source/extensions/filters/common/attributes/attributes.cc

```diff
@@ -0,0 +1,551 @@
+#include "source/common/common/fmt.h"
+#include "source/common/common/lock_guard.h"
+#include "source/common/http/header_map_impl.h"
+
+#include "source/extensions/filters/common/attributes/attributes.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace Attributes {
+
+Http::RegisterCustomInlineHeader<Http::CustomInlineHeaderRegistry::Type::RequestHeaders>
+    referer_handle = Http::CustomHeaders::get().Referer;
+
+MapValue_Entry* ValueUtil::getOrInsert(MapValue* m, absl::string_view key) {
+  if (m == nullptr) {
+    return nullptr;
+  }
+  auto entries = m->mutable_entries();
+  for (auto iter = entries->pointer_begin(); iter < entries->pointer_end(); iter++) {
+    auto entry = *iter;
+    if (entry->has_key() && entry->key().has_string_value() && entry->key().string_value() == key) {
+      return entry;
+    }
+  }
+
+  // add a new entry to entries
+  MapValue_Entry* e = m->add_entries();
+  auto k = e->mutable_key();
+  k->set_string_value(std::string(key));
+
+  return e;
+}
+
+Value* ValueUtil::mapValue(MapValue* m) {
+  Value* val = new Value();
+  val->set_allocated_map_value(m);
+  return val;
+}
+
+Value* ValueUtil::stringValue(const std::string& str) {
+  Value* val = new Value();
+  val->set_string_value(str);
+  return val;
+}
+
+Value* ValueUtil::optionalStringValue(const absl::optional<std::string>& str) {
+  if (str.has_value()) {
+    return ValueUtil::stringValue(str.value());
+  }
+  return ValueUtil::nullValue();
+}
+
+Value* ValueUtil::uint64Value(uint64_t n) {
+  Value* val = new Value();
+  val->set_uint64_value(n);
+  return val;
+}
+
+Value* ValueUtil::boolValue(bool b) {
+  Value* val = new Value();
+  val->set_bool_value(b);
+  return val;
+}
+
+Value* ValueUtil::nullValue() {
+  Value* v = new Value();
+  Value iv;
+  v->set_null_value(iv.null_value());
+  return v;
+}
+
+template <class T> Value* ValueUtil::objectValue(T val) {
+  Value* v = new Value();
+  Any* a = v->mutable_object_value();
+  a->PackFrom(*val);
+
+  return v;
+}
+
+Value Attributes::buildAttributesValue(const std::vector<AttributeId>& attrs) {
+  Value* value = new Value();
+  auto map = value->mutable_map_value();
+
+  for (auto attr : attrs) {
+    RootToken root_token = attr.root();
+    MapValue_Entry* root_entry = ValueUtil::getOrInsert(map, attr.root_name());
+
+    if (!attr.sub()) {
+      Value val = full(root_token);
+      root_entry->set_allocated_value(&val);
+    } else {
+      Value* root_entry_val = root_entry->mutable_value();
+      MapValue* sub_map = root_entry_val->mutable_map_value();
+
+      auto sub_key = attr.sub_name();
+      MapValue_Entry* sub_entry = ValueUtil::getOrInsert(sub_map, *sub_key);
+      Value* sub_val = get(attr);
+
+      sub_entry->set_allocated_value(sub_val);
+    }
+  }
+
+  auto v = *value;
+  return v;
+}
+
+Value* Attributes::get(AttributeId& attr_id) {
+  std::cerr << "getting attr_id: " << attr_id.root_name() << "\n";
```

Comment:
Can you please replace this and other output with ENVOY_LOG macros? You already have this class inheriting from Loggable so ENVOY_LOG(trace, "message {}", parameter) kind of stuff will work.